### PR TITLE
Allow for dict-like argument to Categorical.rename_categories

### LINKED ISSUE #17336: Allow for dict-like argument to Categorical.rename_categories
#### Proposal

I think it would be great if [pandas.core.categorical.Categorical.rename_categories](http://pandas.pydata.org/pandas-docs/version/0.15.2/generated/pandas.core.categorical.Categorical.rename_categories.html) could take a dict-like argument as an alternative to the index-like argument that it currently requires. This would make the syntax more similar to that of [DataFrame.rename](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.rename.html). The proposed change would allow for something like this:

#### Code Sample

```python
>>>> df['categorical_var']
0    apples
1    oranges
2    apples
3    apples

>>>> # rename categories from 'apples' and 'oranges' to 'red fruits' and 'orange fruits'
>>>> df['categorical_var'].rename_categories({ 'apples': 'red fruits', 'oranges': 'orange fruits'})
0    red fruits
1    orange fruits
2    red fruits
3    red fruits
```

#### Why?

**1. Allows the developer to execute the rename without any prior knowledge of the order of the categories.**

This:
```python
>>>> df['categorical_var']
0    apples
1    oranges
2    apples
3    apples

>>>> df['categorical_var'].rename_categories({ 'apples': 'red fruits', 'oranges': 'orange fruits'})
0    red fruits
1    orange fruits
2    red fruits
3    red fruits
```
instead of this:
```python
>>>> df['categorical_var']
0    apples
1    oranges
2    apples
3    apples

>>>> # to figure out current category order
>>>> print(df['categorical_var'].cat.categories)
['apples', 'oranges']

>>>> # rename based on order
>>>> df['categorical_var'].rename_categories(['red fruits', 'orange fruits'])
0    red fruits
1    orange fruits
2    red fruits
3    red fruits
```

**2. Allows the developer to much more easily rename one or a small subset of categories without having to worry about the rest of the category names.**

This:
```python
>>>> df['categorical_var']
0    apples
1    oranges
2    apples
3    apples
4    green fruits
5    yellow fruits
6    blue fruits
7    purple fruits

>>>> df['categorical_var'].rename_categories({ 'apples': 'red fruits', 'oranges': 'orange fruits'})
0    red fruits
1    orange fruits
2    red fruits
3    red fruits
4    green fruits
5    yellow fruits
6    blue fruits
7    purple fruits
```
instead of this:
```python
>>>> df['categorical_var']
0    apples
1    oranges
2    apples
3    apples
4    green fruits
5    yellow fruits
6    blue fruits
7    purple fruits

>>>> # to find full list of categories
>>>> print(df['categorical_var'].cat.categories)
['apples', 'oranges' 'green fruits', 'yellow fruits', 'blue fruits', 'purple fruits']

>>>> # rename must contain full category list
>>>> df['categorical_var'].rename_categories(['red fruits', 'orange fruits', 'green fruits', 'yellow fruits', 'blue fruits', 'purple fruits'])
0    red fruits
1    orange fruits
2    red fruits
3    red fruits
4    green fruits
5    yellow fruits
6    blue fruits
7    purple fruits
```

**3. As I mentioned above, this also makes the syntax more similar to the DataFrame.rename syntax, which I think increases overall usability**

I would be happy to start working on this if others agree that adding the feature makes sense.

### PR DESCRIPTION
**Pull request by alanbato**
- [x] closes #17336
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

`Categorical.rename_categories` can now take a dict as `new_categories` and updates the categories to those found in the mapping.
 It will only change the categories found in the dict, and leave all others the same.
The dict can be smaller or larger, and may or may not contain mappings referencing the old categories.

Have a good day/night! 🐼 🐍

### CODE REVIEW

File: doc/source/whatsnew/v0.21.0.txt [hunk 1]
```diff
@@ -114,7 +114,7 @@ Other Enhancements
 - :func:`pd.read_sas()` now recognizes much more of the most frequently used date (datetime) formats in SAS7BDAT files (:issue:`15871`).
 - :func:`DataFrame.items` and :func:`Series.items` is now present in both Python 2 and 3 and is lazy in all cases (:issue:`13918`, :issue:`17213`)
 - :func:`Styler.where` has been implemented. It is as a convenience for :func:`Styler.applymap` and enables simple DataFrame styling on the Jupyter notebook (:issue:`17474`).
-
+- :func:`Categorical.rename_categories` now accepts a dict-like argument as `new_categories`, and only updates the categories found in that dict. (:issue:`17336`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: can you update / add to the example in categorical.rst as well

File: pandas/core/categorical.py [hunk 2]
```diff
@@ -824,7 +824,11 @@ def rename_categories(self, new_categories, inplace=False):
         """
         inplace = validate_bool_kwarg(inplace, 'inplace')
         cat = self if inplace else self.copy()
-        cat.categories = new_categories
+        if isinstance(new_categories, dict):
```
Comment (jreback) [inline pandas/core/categorical.py]: use is_dict_like, no need to do this as an if

```
if is_dict_like(new_categories):
    new_categories = [new_categories.get(item, item).....]
cat.categories = new_categories
```

File: pandas/tests/test_categorical.py [hunk 3]
```diff
@@ -983,6 +983,34 @@ def test_rename_categories(self):
...
+        cat = pd.Categorical(['a', 'b', 'c', 'd'])
+        res = cat.rename_categories({'a': 4, 'b': 3, 'c': 2, 'd': 1})
+        expected = Index([4, 3, 2, 1])
+        tm.assert_index_equal(res.categories, expected)
+        # Test for inplace
+        res = cat.rename_categories({'a': 4, 'b': 3, 'c': 2, 'd': 1},
+                                    inplace=True)
+        assert res is None
+        tm.assert_index_equal(cat.categories, expected)
+        # Test for dicts of smaller length
```
Comment (jreback) [inline pandas/tests/test_categorical.py]: blank lines between sub-tests

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. some small comments. pls rebase / ping on green.

**Comment by TomAugspurger** [comment]
Don't worry about the travis failures. They're unrelated and have been fixed. +1 once you change the `if isinstance(...)` to use `is_dict_like`.

File: doc/source/whatsnew/v0.21.0.txt [hunk 4]
```diff
@@ -115,6 +115,7 @@ Other Enhancements
 - :func:`DataFrame.items` and :func:`Series.items` is now present in both Python 2 and 3 and is lazy in all cases (:issue:`13918`, :issue:`17213`)
 - :func:`Styler.where` has been implemented. It is as a convenience for :func:`Styler.applymap` and enables simple DataFrame styling on the Jupyter notebook (:issue:`17474`).
 - :func:`MultiIndex.is_monotonic_decreasing` has been implemented.  Previously returned ``False`` in all cases. (:issue:`16554`)
+- :func:`Categorical.rename_categories` now accepts a dict-like argument as `new_categories`, and only updates the categories found in that dict. (:issue:`17336`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.21.0.txt]: Nit: remove the comma before "and only updates"

File: pandas/core/categorical.py [hunk 5]
```diff
@@ -824,7 +825,11 @@ def rename_categories(self, new_categories, inplace=False):
         """
         inplace = validate_bool_kwarg(inplace, 'inplace')
         cat = self if inplace else self.copy()
-        cat.categories = new_categories
+        if is_dict_like(new_categories):
```
Comment (gfyoung) [inline pandas/core/categorical.py]: Nit: newline above this one.

File: pandas/tests/test_categorical.py [hunk 6]
```diff
@@ -983,6 +983,35 @@ def test_rename_categories(self):
         with pytest.raises(ValueError):
             cat.rename_categories([1, 2])
 
+    def test_rename_categories_dict(self):
+        # GH 17336
+        cat = pd.Categorical(['a', 'b', 'c', 'd'])
+        res = cat.rename_categories({'a': 4, 'b': 3, 'c': 2, 'd': 1})
+        expected = Index([4, 3, 2, 1])
+        tm.assert_index_equal(res.categories, expected)
+        # Test for inplace
```
Comment (gfyoung) [inline pandas/tests/test_categorical.py]: Nit: newline above this one.

File: pandas/tests/test_categorical.py [hunk 7]
```diff
@@ -983,6 +983,35 @@ def test_rename_categories(self):
...
+        res = cat.rename_categories({'a': 4, 'b': 3, 'c': 2, 'd': 1},
+                                    inplace=True)
+        assert res is None
+
+        tm.assert_index_equal(cat.categories, expected)
+        # Test for dicts of smaller length
```
Comment (gfyoung) [inline pandas/tests/test_categorical.py]: Nit: newline above this one.

File: pandas/tests/test_categorical.py [hunk 8]
```diff
@@ -983,6 +983,35 @@ def test_rename_categories(self):
...
+        cat = pd.Categorical(['a', 'b', 'c', 'd'])
+        res = cat.rename_categories({'a': 4, 'b': 3, 'c': 2, 'd': 1})
+        expected = Index([4, 3, 2, 1])
+        tm.assert_index_equal(res.categories, expected)
+        # Test for inplace
+        res = cat.rename_categories({'a': 4, 'b': 3, 'c': 2, 'd': 1},
+                                    inplace=True)
+        assert res is None
+
+        tm.assert_index_equal(cat.categories, expected)
```
Comment (gfyoung) [inline pandas/tests/test_categorical.py]: Nit: remove newline above this one.

File: pandas/tests/test_categorical.py [hunk 9]
```diff
@@ -983,6 +983,35 @@ def test_rename_categories(self):
...
+        # Test for dicts of smaller length
+        cat = pd.Categorical(['a', 'b', 'c', 'd'])
+        res = cat.rename_categories({'a': 1, 'c': 3})
+        expected = Index([1, 'b', 3, 'd'])
+        tm.assert_index_equal(res.categories, expected)
+        # Test for dicts with bigger length
```
Comment (gfyoung) [inline pandas/tests/test_categorical.py]: Nit: newline above this one.

File: pandas/tests/test_categorical.py [hunk 10]
```diff
@@ -983,6 +983,35 @@ def test_rename_categories(self):
...
+        cat = pd.Categorical(['a', 'b', 'c', 'd'])
+        res = cat.rename_categories({'a': 1, 'b': 2, 'c': 3,
+                                     'd': 4, 'e': 5, 'f': 6})
+        expected = Index([1, 2, 3, 4])
+        tm.assert_index_equal(res.categories, expected)
+        # Test for dicts with no items from old categories
```
Comment (gfyoung) [inline pandas/tests/test_categorical.py]: Nit: newline above this one.

File: pandas/tests/test_categorical.py [hunk 11]
```diff
@@ -983,6 +983,35 @@ def test_rename_categories(self):
...
+        # Test for inplace
+        res = cat.rename_categories({'a': 4, 'b': 3, 'c': 2, 'd': 1},
+                                    inplace=True)
+        assert res is None
+
+        tm.assert_index_equal(cat.categories, expected)
+        # Test for dicts of smaller length
+        cat = pd.Categorical(['a', 'b', 'c', 'd'])
+        res = cat.rename_categories({'a': 1, 'c': 3})
+        expected = Index([1, 'b', 3, 'd'])
```
Comment (gfyoung) [inline pandas/tests/test_categorical.py]: Nit: newline above this one.

File: pandas/tests/test_categorical.py [hunk 12]
```diff
@@ -983,6 +983,35 @@ def test_rename_categories(self):
...
+        # Test for dicts with bigger length
+        cat = pd.Categorical(['a', 'b', 'c', 'd'])
+        res = cat.rename_categories({'a': 1, 'b': 2, 'c': 3,
+                                     'd': 4, 'e': 5, 'f': 6})
+        expected = Index([1, 2, 3, 4])
+        tm.assert_index_equal(res.categories, expected)
+        # Test for dicts with no items from old categories
+        cat = pd.Categorical(['a', 'b', 'c', 'd'])
+        res = cat.rename_categories({'f': 1, 'g': 3})
+        expected = Index(['a', 'b', 'c', 'd'])
```
Comment (gfyoung) [inline pandas/tests/test_categorical.py]: Nit: newline above this one.

**Comment by gfyoung** [comment]
@alanbato : LGTM as well.  Bunch of nit-picking to enhance readability is all I've got above.

**Comment by alanbato** [comment]
Done @gfyoung. Do we wait for green again?

**Comment by gfyoung** [comment]
@alanbato : Yes, ping us when it's green.

**Review by gfyoung — APPROVED** [review]

**Comment by alanbato** [comment]
@TomAugspurger @gfyoung @jreback Ping! 🌵✔️

**Comment by gfyoung** [comment]
Nice!  I'll let @jreback look this over and merge just to double check.

**Review by jreback — APPROVED** [review]
lgtm. merge away

File: pandas/core/categorical.py [hunk 13]
```diff
@@ -824,7 +825,12 @@ def rename_categories(self, new_categories, inplace=False):
         """
         inplace = validate_bool_kwarg(inplace, 'inplace')
         cat = self if inplace else self.copy()
-        cat.categories = new_categories
+
+        if is_dict_like(new_categories):
+            cat.categories = [new_categories.get(item, item)
```
Comment (jreback) [inline pandas/core/categorical.py]: actually make sure the doc string is updated here as well (maybe add a versionchanged tag)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/categorical.py [hunk 14]
```diff
@@ -792,19 +793,20 @@ def set_categories(self, new_categories, ordered=None, rename=False,
...
-            If the new categories do not have the same number of items than the
-            current categories or do not validate as categories
+            If new categories are list-like and do not have the same number of
+            items than the current categories or do not validate as categories
 
         Parameters
         ----------
-        new_categories : Index-like
+        new_categories : Index-like or Dict-like
            The renamed categories.
```
Comment (jreback) [inline pandas/core/categorical.py]: ``dict-like (>=0.21.0)``

**Comment by alanbato** [comment]
@jreback 🌵 ✔️

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @alanbato nice patch!

**Pull request merged**